Updated aht20 driver to use thread safe api from idf>=5.4.

jeetrohan opened the pull request:
<!--
- Read and understand the project style guidelines (`CONTRIBUTION.md`).
- For Work In Progress Pull Requests, please use the Draft PR feature. See https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
- For a timely review/response, please avoid force-pushing additional commits if your PR has already received reviews or comments.
- Include screenshots for any CLI or UI changes.
- Keep PRs as small as possible; large PRs are difficult to review.
-->

## Description

<!--
- Please include a summary of the changes and the related issue.
- Also include the motivation (why this change) and context.
-->

<!-- 
- If you want to insert images (screenshots, diagrams, etc.), please format them:
    Bad link to the image (not formatted):   ![image](https://github.com/user-attachments/assets/ad3383cd-8f38-4d06-9ecf-3305e299307d)~~
    Good link to the image (formatted):       <img src="https://github.com/user-attachments/assets/ad3383cd-8f38-4d06-9ecf-3305e299307d" width=500px>
-->

## Related

<!--
- Use this format to link issue numbers: Fixes #123 - https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue-using-a-keyword
- Mention any other PRs related to this one.
- If there is related documentation, add the link here.
- If there is a public chat where changes in this PR were initiated, you can include the link here.
-->

## Testing

<!--
- Explain how you tested your change or new feature.
- If you tested changes in a clean environment (Docker container, new virtualenv, fresh Virtual Machine), state it here.
-->

---

## Checklist

Before submitting a Pull Request, please ensure the following:

- [ ] 🚨 This PR does not introduce breaking changes.
- [ ] All CI checks (GH Actions) pass.
- [ ] Documentation is updated as needed.
- [ ] Tests are updated or added as necessary.
- [ ] Code is well-commented, especially in complex areas.
- [ ] Git history is clean — commits are squashed to the minimum necessary.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-iot-solution?pullRequest=503) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-iot-solution?pullRequest=503) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-iot-solution?pullRequest=503) it.</sub>

leeebo commented:
Hi @jeetrohan , thanks for your contribution!

YanKE01 commented on components/sensors/humiture/aht20/idf_component.yml:
```diff
@@ -1,4 +1,23 @@
+    -path: ../../../../examples/sensors/aht20_demo
+
+dependencies:
+  idf: ">=5.4"
```
Forcibly restricting the component in version 5.4 will make it difficult for many existing users of this component to accept. Perhaps the [i2c_bus ](https://components.espressif.com/components/espressif/i2c_bus/versions/1.4.0)component can be used to manage the entire I2C, since i2c_bus will automatically switch to the esp_driver_i2c driver in version 5.3 and above.

jeetrohan replied:
There are two reasons I had originally decided to use idf>=5.4.
1. The new I2C APIs introduced in this version are thread-safe, at least when multiple threads are not accessing the same device/sensor. Basically multiple devices can be accessed safely on same I2C bus, without contention from different threads. 
The new I2C APIs, if I remember correctly, introduced in idf-5.2 are incompatible with older ones(idf<5.2), so it should not affect much the users who are using the new APIs, as they are likely to update sooner than later. 

2.I had assumed that using the Generic I2C driver APIs provide more flexibility to this driver, even when used without this repository (such as component registry, though I might be uninformed on this part, as component registry resolves all the dependencies). 

However, considering only this repository, and taking into account your suggestion.. I've concluded that there are two feasible things that can be done...:

1. I simply do as have been suggested. Replace the driver to i2c_bus component. This will effectively strip the thread-safety for any users, not using idf>=5.4(A note can be placed in this regard in README section).
The dependency in this case, If my understanding is correct, will change to idf>=4.3.
 
2. We release two versions of this driver. One with the i2c_bus component (v1.1.0) and other with esp_driver_i2c (v2.0.0) as the older i2c driver and its APIs are just a legacy support feature.
A relevant note from idf 5.2.5 documentation (& still in idf.5.4.1) : "Please keep in mind that the legacy driver is now deprecated and will be removed in future."
This was the reason to refer this as v2.0.0; to indicate that it is not meant to support legacy driver.

While you finalize the decision.. I'll prepare both the versions.

YanKE01 replied:
Hi, @jeetrohan , in the current repository, we are standardizing the use of `i2c_bus `to manage I2C-type sensors, so I believe using `i2c_bus `would be a better decision. Moreover, `i2c_bus `was designed with thread safety in mind from the beginning, so users don't need to worry about concurrency issues.

As for the AHT20, which is a commonly used component, its upgrade cannot simply switch to the latest version of the IDF (ESP-IDF), because we need to consider the versions being used by current users. If we force the use of the latest IDF, users who are still on versions earlier than IDF 5.4 will no longer be able to use the AHT20 component.

jeetrohan replied:
Hi @YanKE01 
I've updated the most recent submission to use the i2c_bus.
Yes, I understand the support for previous idf is still valid.

YanKE01 commented on components/sensors/humiture/aht20/test_app/main/aht20_test_app.c:
```diff
@@ -0,0 +1,118 @@
+
+    printf("AHT20 deinitialized\n");
+}
+
```
Memory leak checks have not been performed. It might be better to make modifications based on the original `test_apps`, since CI-related issues are also involved.

jeetrohan replied:
I've updated this locally, waiting for your decision on the other recommendation.

YanKE01 commented on components/sensors/humiture/aht20/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 # ChangeLog
 
+## v2.0.0 (2025-04-28
```
It is recommended to update the version to `1.1.0`. Additionally, one side of the parenthesis is missing.

jeetrohan replied:
Fixed the version.

YanKE01 commented on components/sensors/humiture/aht20/README.md:
```diff
@@ -1,32 +1,91 @@
 
-# Component: AHT20
-I2C driver and definition of AHT20 humidity and temperature sensor. 
+#Features
```
```suggestion
# Features
```

jeetrohan replied:
I've updated this locally, waiting for your decision on the other recommendation.

YanKE01 reviewed (commented):
Hi, @jeetrohan , thanks for your contribution!  Before making a commit, you can install pre-commit to ensure that your code style is consistent with the repository.

```
pip install pre-commit
pre-commit install --allow-missing-config -t pre-commit -t commit-msg
```

jeetrohan commented:
> Hi, @jeetrohan , thanks for your contribution! Before making a commit, you can install pre-commit to ensure that your code style is consistent with the repository.
> 
> ```
> pip install pre-commit
> pre-commit install --allow-missing-config -t pre-commit -t commit-msg
> ```

hi @YanKE01 
I'm now having this error which is not letting me make commits and I'm unable to understand why (my idf_component file has version). Can you help me fix this:


Check copyright notices..................................................Passed
file contents sorter.................................(no files to check)Skipped
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
check that executables have shebangs.................(no files to check)Skipped
check that scripts with shebangs are executable..........................Passed
mixed line ending........................................................Passed
fix double quoted strings............................(no files to check)Skipped
Do not use more than one slash in the branch name........................Passed
Do not use uppercase letters in the branch name..........................Passed
astyle formatter.........................................................Passed
Check File Permissions...............................(no files to check)Skipped
Validate executable-list.txt.............................................Passed

Update versions in readme................................................Failed
- hook id: release_versions
- exit code: 1

Traceback (most recent call last):
  File "my-esp-iot-solution/tools/release_versions.py", line 29, in <module>
    components = get_components_versions(os.path.join(os.path.dirname(__file__), '../components'))
  File "my-esp-iot-solution/tools/release_versions.py", line 25, in get_components_versions
    component_version = data['version']
KeyError: 'version'

codespell................................................................Passed

jeetrohan commented:
Hi @YanKE01 
I've refactored all the recommendations you made regarding this driver.
I had to do a hard reset because of previous commits without hook. Apparently that was the reason I was unable to make commits after installing pre-commit hook and soft reset didn't helped.

If you still find anything that needs to be fixed, let me know.
I'll be releasing the aforementioned v2.0.0 as v1.*.* something in my component registry, so that if ever needed....

Also, I haven't merged this with Sensor Hub, because..... :zipper_mouth_face: 
Let's just say, I do feel like Sensor Hub can do more.

Thanks for pointing me out to use the i2c_bus component, I dislike it for my own reasons, but it was helpful in its own ways.
Also, is there a unified interface for SPI and I2C bus? I have tested a sensor (will be testing another probably soon enough) with both buses.
I wanted to contribute those too to this repository. If its not there then I'll simply make shift the things.

jeetrohan commented:
Hi, @YanKE01 & @leeebo 
Could you confirm me if the current submission is acceptable? or there is a need for further change?

YanKE01 commented:
Hi, @jeetrohan . Thank you very much for your contribution. The `sensor_hub `is still under development. Could you list some of the current issues and features you would like to see added? This will help us improve `sensor_hub `more effectively.

Additionally, the design principles of `i2c_bus `and `spi_bus `are the same, and both have taken the resource contention issues you mentioned into account. If you have any questions about `i2c_bus `or `spi_bus`, please let us know so we can further improve them.

jeetrohan commented:
> Hi, @jeetrohan . Thank you very much for your contribution. The `sensor_hub `is still under development. Could you list some of the current issues and features you would like to see added? This will help us improve `sensor_hub `more effectively.

Hi @YanKE01, 
First **TLDR** : It is currently feature rich as it is (to a good extent, if not wholly) and I would like to contribute myself for many things to the implementation/framework itself.
 
This will become a very long rant if I will go on explaining everything. Even without details :

I'm not going to talk about features because, in my humble opinion, there are plenty of helpful features in current implementation. Additionally, no matter the amount of features, somebody can always come up with some niche but kind of essential feature.

What I'm going to talk about is, its structure or the framework itself.

1. **In current implementation as it is:**

- I remember seeing a small bug/resource management issue in the code, among many other things, which I'll fix and open PR for, myself. But I vaguely remember some other things too.
- There are few other things which I need to confirm by using it myself (I saw the code for whole thing and how it is connected and could cause issues but not certain at the moment). 

2.  **Things I would like to contribute, myself:**

-  Current implementation, as I see it is a lot of good things put together with good intentions... but.. I don't see where it is going.. I mean I see it... but.. this is something which would require a more formal form of discussion (to you this is, not to me, I'm not the part of development team, just a contributor.)
- In addition to that.. there are some little things which I would like to add myself.
However, this is the kind of work which makes more sense when I'm part of team. I don't know how you would evaluate my current work, but mine is : sloppy(my evaluation of my work I've submitted up until in any form to esp-idf). And that's because there's nothing associated in it for me. I would rather like to do sincere work.

Where is this coming from??
I have been using the esp-idf for quite sometime and there are little things, that keep feeling like why they are not there..? I understand there is a lot of workload on all the teams, because too many things are being pursued.

**What I can do at the moment for certain:**
I can definitely contribute 2 more sensor driver for the Hub (if acceptable). Both tested on hardware. (I'm not certain about contributing a few others, will need time. One of these is not the kind where a simple implementation of data sheet does it.) 

Why the delay in some things?
Currently I'm trying to implement a few other things related to BLE (and also another RTOS).
Once I'm done, I'll update with more certainty.

jeetrohan commented:
> Additionally, the design principles of `i2c_bus `and `spi_bus `are the same, and both have taken the resource contention issues you mentioned into account. If you have any questions about `i2c_bus `or `spi_bus`, please let us know so we can further improve them.

I've not checked the code... but...
Since the i2c_bus internally itself uses the underlying esp_driver_i2c, would it not be a better idea to remove the resource lock used in i2c_bus for idf>=5.4.
It would be unnecessary to double lock the resource.. once in i2c_bus and then in esp_driver_i2c. 
Unless.... the lock in i2c_bus  guarantees thread safety even when the same resource/sensor on same bus is being accessed by two different threads. (my understanding of esp_driver_i2c from documentation is that it doesn't provides this access security.)

jeetrohan commented:
Hi @YanKE01 ,
If you want me to add Sensor Hub support in this driver I can do that too... But I won't be testing it (for reasons I've mentioned in other reply.) and this is also the reason I dropped the idea of doing it. 
According to documents and other sensors.. I need to just add a few more lines.. but from my perspective.. adding those lines is additional testing, and one more example, basically few hours. 
So, if you're ok.. with having a few more lines added.. let me know.

And if there is any other thing to finalize this contribution, please let me know. 
Thanks.

jeetrohan commented:
> You can see from the commit history of `i2c_bus` that it predates the current `esp_driver_i2c` driver by quite a bit. Moreover, the design philosophy of `esp_driver_i2c` shares similarities with `i2c_bus`, such as separating the bus and device, and ensuring thread safety.
> 
> As for your question about why `esp_driver_i2c` now provides thread-safe bus operation functions, yet `i2c_bus` still performs additional locking — the reason is that, at the time `i2c_bus` was made compatible with `esp_driver_i2c`, the latter did not yet guarantee thread-safe signaling for bus operations like it does now. You can refer to the documentation of `esp_driver_i2c` from version `5.3.3` to verify this.
> 
> `ESP-IDF` drivers are constantly being improved, but as a component, I believe compatibility should be the top priority — because you can't expect everyone to always stay updated with the latest IDF master branch.

@YanKE01,
I apologize.. I suppose I didn't put it in the right way...
Allow me to put it more properly...

1. i2c_bus is meant to provide, yet another layer of abstraction, because currently the lifecycle of previous idf versions is going to exist until at least 25Jan2027 (esp-idf-5.3).
2. i2c_bus is meant to provide necessary features not available in older idf, and as such serves essential purpose.
3. It separates legacy driver from new driver by using this... #if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 3, 0).

4. _My suggestion was... if possible._... the locking mechanism in i2c_bus should use something similar to avoid putting additional locking mechanism in esp-idf>=5.4 to avoid double locking of resources.....(as I said I haven't looked inside the code, but I'm assuming this should be possible. Though it may make it a bit stuffy. Probably, also it will not make any significant difference in performance.. unless very time critical tasks are being performed, but at least will not add another layer during runtime).

5. _But... the above suggestion has no value, if....._ the lock in i2c_bus guarantees thread safety even when the same resource/sensor on same bus is being accessed by two different threads. (my understanding of esp_driver_i2c from documentation is that it doesn't provides this access security.)

This will ensure both compatibility with previous versions.. without adding another layer of locking mechanism on an existing one (provided it is same as the one in esp-idf=5.4).

YanKE01 commented:
Hi, @jeetrohan.  I still recommend that you use `i2c_bus` to update the `aht20` component, as it is compatible with different versions of the IDF and does not force users to switch to the latest IDF version. If you still prefer to use the `esp_driver_i2c` API, you can try submitting it to your own space in the component manager. You can refer to the relevant [submission documentation](https://docs.espressif.com/projects/idf-component-manager/en/latest/getting_started/index.html).

jeetrohan commented:
![Screenshot from 2025-05-08 22-35-26](https://github.com/user-attachments/assets/81a5c9d1-5ab0-48fd-9f17-0aa00e66873e)
> Hi, @jeetrohan. I still recommend that you use `i2c_bus` to update the `aht20` component, as it is compatible with different versions of the IDF and does not force users to switch to the latest IDF version. If you still prefer to use the `esp_driver_i2c` API, you can try submitting it to your own space in the component manager. You can refer to the relevant [submission documentation](https://docs.espressif.com/projects/idf-component-manager/en/latest/getting_started/index.html).

@YanKE01 
I totally agree with your sentiment.


Please check the commit log from from last week's submission.

I've already updated this driver to use i2c_bus and I've informed you regarding the same.
![Screenshot from 2025-05-08 22-35-26](https://github.com/user-attachments/assets/7c49ce3f-1763-4c36-aa39-0ebde0db3f67)

YanKE01 commented on components/sensors/humiture/aht20/aht20.c:
```diff
@@ -1,160 +1,308 @@
-    ESP_RETURN_ON_FALSE(handle, ESP_ERR_INVALID_ARG, TAG, "invalid device handle pointer");
+    ESP_RETURN_ON_FALSE((aht20_handle != NULL), ESP_ERR_INVALID_ARG, s_TAG, "empty handle, initialize AHT20 handle");
+
+    vTaskDelay(20 / portTICK_PERIOD_MS); //time for AHT20 SCL to stabilize
```
As I²C is a master-driven protocol, I'm a bit unsure about the necessity of waiting for SCL to stabilize at this point. Could you clarify why this delay is needed and how the specific duration was chosen?

jeetrohan replied:
@YanKE01 still waiting for your response.
Was my reply sufficient enough or should I remove this delay?

jeetrohan replied:
@YanKE01 
Thank you for the suggestion.
Please consider the following and if yous still believe, it should be configurable, please provide me with exact range and default value that you believe is suitable, adding a simple config option is not a problem.
Although, I believe it should be left as it is.

From the AHT20 datasheet:
![image](https://github.com/user-attachments/assets/40088650-d745-4343-b608-45bbcc391b49)

YanKE01 replied:
Let's just follow the specifications in the datasheet—it doesn't seem to require any additional configuration.

YanKE01 commented on components/sensors/humiture/aht20/test_apps/main/Kconfig.projbuild:
```diff
@@ -2,13 +2,13 @@ menu "Example Configuration"
     config I2C_MASTER_SCL
         int "SCL GPIO Num"
-        default 40
+        default 22
```
Since this is a Kconfig option, I think it's fine to leave it as is unless there's a strong reason to change it — users can always configure it to the IO they need.

jeetrohan replied:
fixed.
I had just updated it with values I used on esp32.
Had replaced with local files during updated, rather than editing.

YanKE01 commented on examples/sensors/aht20_demo/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# The following five lines of boilerplate have to be in your project's
```
The usage of AHT20 is already clearly explained in the component's README. @leeebo , do you think it's still necessary to write an additional example?

jeetrohan replied:
@YanKE01,
 Sorry.. I didn't noticed the question was directed not at me, until I was about to comment..
But since.. I've.. 
consider this as my opinion or suggestion to both, you and @leeebo.

Of course, I will remove the example from my end, if that's the conclusion you two reach to.


Many, if not most, developers will agree with your sentiment; maybe I'm the odd one out.
 
Allow me to rephrase your question:
"There's a perfectly fine instruction manual so why the demo?"

I'll answer you as a typical layman user, in the most laymen language I can :

Note: Laymen in this context is a user who knows his way around C language, IDEs and has experience of working with micro-controllers but has zero interest in understanding the underlying framework used in esp32 ecosystem.

"I just want to use this sensor, to get some readings.. I don't want to bother to implement it myself.. and definitely not interested in how the readings are being collected by underlying framework... I just want to get the readings and use them for whatever insidious(humor, not serious) purposes I have. 
I don't want to implement nothing (intentionally written instead of anything)."

In more simpler words.. Its for people used to Arduino, fire and forget.

jeetrohan replied:
@leeebo I'm still waiting for your response. I will use that as guideline for the next two sensors I'll be contributing to this repository.(drivers already tested on hardware).

jeetrohan commented on components/sensors/humiture/aht20/CHANGELOG.md:
```diff
@@ -1,5 +1,8 @@
 # ChangeLog
 
+## v1.1.1 (2025-05-02)
+* replace the i2c interface with i2c_bus
```
fixed.

jeetrohan commented on components/sensors/humiture/aht20/test_apps/CMakeLists.txt:
```diff
@@ -1,8 +1,7 @@
 
-set(EXTRA_COMPONENT_DIRS "$ENV{IDF_PATH}/tools/unit-test-app/components"
-                         "../../aht20")
+set(COMPONENTS main)
```
fixed.

jeetrohan commented on components/sensors/humiture/aht20/README.md:
```diff
@@ -1,32 +1,90 @@
-[![Component Registry](https://components.espressif.com/components/espressif/aht20/badge.svg)](https://components.espressif.com/components/espressif/aht20)
+# aht20
```
fixed.

jeetrohan commented on components/sensors/humiture/aht20/README.md:
```diff
@@ -1,32 +1,90 @@
+
+    Follow the example to learn how to initialize the driver and read the sensor data.
+
+    All public APIs are documented in aht20.h.
```
fixed.

jeetrohan commented on components/sensors/humiture/aht20/README.md:
```diff
@@ -1,32 +1,90 @@
-    aht20_new_sensor(&i2c_conf, &handle);
+    //create a AHT20 device object and receive a device handle for it
+    // my_i2c_bus_handle is a preintialized i2c_bus_handle_t object
+    aht20_handle_t aht20_handle =  aht20_create( my_i2c_bus_handle, AHT20_ADDRESS_LOW ); //addresses are in aht.h
```
fixed.

jeetrohan commented on components/sensors/humiture/aht20/aht20.c:
```diff
@@ -1,160 +1,308 @@
+    return ESP_OK;
+}
+
+static esp_err_t aht_JH_Reset_REG(aht20_handle_t aht20_handle, uint8_t addr)
```
fixed.

jeetrohan commented on components/sensors/humiture/aht20/test_apps/CMakeLists.txt:
```diff
@@ -1,8 +1,7 @@
 # The following lines of boilerplate have to be in your project's CMakeLists
 # in this exact order for cmake to work correctly
-cmake_minimum_required(VERSION 3.5)
+cmake_minimum_required(VERSION 3.16)
```
fixed.
But,
I had replaced the original files with my local files rather than editing. 
Please check the compilation after this change with corresponding CMake version.

jeetrohan commented on components/sensors/humiture/aht20/README.md:
```diff
@@ -1,32 +1,90 @@
+
+
+
+However, following are the general guiedlines.
```
fixed.
However, I would like to inform you that..
It was perfectly right to use 'however' in above context... because it not always represents inconsistency.
It is also used for adding a comment to what have been just said, although something is true.

It was supposed to reflect that .... although the user should follow the example.. but for those who don't want to.. here are are basic guidelines.

jeetrohan commented on components/sensors/humiture/aht20/aht20.c:
```diff
@@ -1,160 +1,308 @@
+    return ESP_OK;
+}
+
+esp_err_t aht20_read_humiture(aht20_handle_t aht20_handle)
```
This... will be a long answer... and one without TLDR.

There are two questions here....

1. Why provide aht20_read_humiture() to read at once both temperature and humidity when there are separate APIs aht20_read_humidity() and aht20_read_temperature().
2. And even if implement it ... why the way it has been?

So the answer to 1. : 
              AHT20 datasheet indicate/directs that readings should be taken at the interval of minimum 2 seconds.
 This means user should wait for at least 2 seconds in between calling aht20_read_humidity() and aht20_read_temperature(); as both of these function are equivalent of a single sensor data read. 
However, it is somewhat inconvenient or at least not desired to have a delay between readings for associated parameters.. although unlikely that there will be much fluctuations in value, within 2 seconds. But still unnecessary when user wants to read both, because every read provides raw values for both.
If my memory serves correctly, some resources indicated that reading faster than 2 seconds may damage hardware, but I'm not sure how reliable this information is.

And answer to 2. :
              I agree, its unconventional to put results in handle, at least in ESP ecosystem. But consider this just as a handy feature.. providing user with more convenient access to the data they are most likely to access with one thing that they will definitely need to access any call to sensor.
 But why as a structure and not simply float variable? Again convenience to user in passing packed data.
 
 
 Now what could be done.. :

- Completely remove aht20_read_humiture()
- Keep aht20_read_humiture(), but, instead of structure use simple float variables in handle.
- Keep aht20_read_humiture(), but return the structure instead of storing in handle.
- Keep the aht20_read_humiture() as it is.

jeetrohan replied:
@YanKE01 is my response sufficient enough to explain the purpose? 
or should I remove this function altogether?
Although, I've come to understand your question better when I delved again in Sensor_Hub implementation. This would appear as redundant.  
However, the implementation of aht20 drive needs to be tweaked a bit in order to display its functionality in that case.

YanKE01 replied:
I'm not very certain myself. I feel like this is a feature I would hardly ever come across.

jeetrohan replied:
@YanKE01 If you feel like the feature has no particular use I don't mind dropping it.. but.. please keep the following in consideration when making a decision.

> **AHT20 datasheet indicate/directs that readings should be taken at the interval of minimum 2 seconds.**
> **This means user should wait for at least 2 seconds in between calling aht20_read_humidity() and aht20_read_temperature(); as both of these function are equivalent of a single sensor data read.**
> However, it is somewhat inconvenient or at least not desired to have a delay between readings for associated parameters.. although unlikely that there will be much fluctuations in value, within 2 seconds. But still unnecessary when user wants to read both, because every read provides raw values for both.

and....

> If my memory serves correctly, some resources indicated that reading faster than 2 seconds may damage hardware, but I'm not sure how reliable this information is.

jeetrohan replied:
> I'm not very certain myself. I feel like this is a feature I would hardly ever come across.

@YanKE01 please also clarify if you're talking about .. 
just this... aht20_read_humiture() 
or
adding the results in handle or both?

YanKE01 commented:
I noticed that in test_apps, the I2C bus handle is created via `esp_driver_i2c` rather than the `i2c_bus` framework. Then, this handle is directly passed to `i2c_bus_device_create` in aht20_create, which is not valid because the two handles do not point to the same internal data structure.

```shell
(1)     "AHT20 Sensor" [aht20][sensor]

Enter test for running.
1
Running AHT20 Sensor...
Requesting I2C bus handle
I2C bus handle acquired
I (1640) AHT20: adding aht20 as device to bus

E (1640) i2c_bus: /home/yanke/project/esp-iot-solution/components/i2c_bus/i2c_bus_v2.c:185 (i2c_bus_device_create):i2c_bus has not inited
E (1650) AHT20: aht20_create(290): unable to create device

E (1650) i2c_bus: /home/yanke/project/esp-iot-solution/components/i2c_bus/i2c_bus_v2.c:214 (i2c_bus_device_delete):Null Device Handle
MALLOC_CAP_8BIT: Before 388860 bytes free, After 388756 bytes free (delta -104)
MALLOC_CAP_32BIT: Before 388860 bytes free, After 388756 bytes free (delta -104)
./main/aht20_test_app.c:123:AHT20 Sensor:PASS
Test ran in 58ms
```

jeetrohan commented:
> I noticed that in test_apps, the I2C bus handle is created via `esp_driver_i2c` rather than the `i2c_bus` framework. Then, this handle is directly passed to `i2c_bus_device_create` in aht20_create, which is not valid because the two handles do not point to the same internal data structure.
> 
> ```shell
> (1)     "AHT20 Sensor" [aht20][sensor]
> 
> Enter test for running.
> 1
> Running AHT20 Sensor...
> Requesting I2C bus handle
> I2C bus handle acquired
> I (1640) AHT20: adding aht20 as device to bus
> 
> E (1640) i2c_bus: /home/yanke/project/esp-iot-solution/components/i2c_bus/i2c_bus_v2.c:185 (i2c_bus_device_create):i2c_bus has not inited
> E (1650) AHT20: aht20_create(290): unable to create device
> 
> E (1650) i2c_bus: /home/yanke/project/esp-iot-solution/components/i2c_bus/i2c_bus_v2.c:214 (i2c_bus_device_delete):Null Device Handle
> MALLOC_CAP_8BIT: Before 388860 bytes free, After 388756 bytes free (delta -104)
> MALLOC_CAP_32BIT: Before 388860 bytes free, After 388756 bytes free (delta -104)
> ./main/aht20_test_app.c:123:AHT20 Sensor:PASS
> Test ran in 58ms
> ```

@YanKE01 
It was already fixed locally yesterday, your response though not updated yet, when....
I noticed this while I was trying to add the Sensor Hub Functionality and also add a unit test for that... however, I'm facing a bit of problem with that.. would you like to help me with that?

or 

can I just add the Sensor Hub relevant code and update it as it is, without the unit test?

YanKE01 commented:
> > I noticed that in test_apps, the I2C bus handle is created via `esp_driver_i2c` rather than the `i2c_bus` framework. Then, this handle is directly passed to `i2c_bus_device_create` in aht20_create, which is not valid because the two handles do not point to the same internal data structure.
> > ```shell
> > (1)     "AHT20 Sensor" [aht20][sensor]
> > 
> > Enter test for running.
> > 1
> > Running AHT20 Sensor...
> > Requesting I2C bus handle
> > I2C bus handle acquired
> > I (1640) AHT20: adding aht20 as device to bus
> > 
> > E (1640) i2c_bus: /home/yanke/project/esp-iot-solution/components/i2c_bus/i2c_bus_v2.c:185 (i2c_bus_device_create):i2c_bus has not inited
> > E (1650) AHT20: aht20_create(290): unable to create device
> > 
> > E (1650) i2c_bus: /home/yanke/project/esp-iot-solution/components/i2c_bus/i2c_bus_v2.c:214 (i2c_bus_device_delete):Null Device Handle
> > MALLOC_CAP_8BIT: Before 388860 bytes free, After 388756 bytes free (delta -104)
> > MALLOC_CAP_32BIT: Before 388860 bytes free, After 388756 bytes free (delta -104)
> > ./main/aht20_test_app.c:123:AHT20 Sensor:PASS
> > Test ran in 58ms
> > ```
> 
> @YanKE01 It was already fixed locally yesterday, when.... I noticed this while I was trying to add the Sensor Hub Functionality and also add a unit test for that... however, I'm facing a bit of problem with that.. would you like to help me with that?
> 
> or
> 
> can I just add the Sensor Hub relevant code and update it as it is, without the unit test?

What specific issues have you encountered?

jeetrohan commented:
> > @YanKE01 It was already fixed locally yesterday, when.... I noticed this while I was trying to add the Sensor Hub Functionality and also add a unit test for that... however, I'm facing a bit of problem with that.. would you like to help me with that?
> > or
> > can I just add the Sensor Hub relevant code and update it as it is, without the unit test?
> 
> What specific issues have you encountered?

This is the issue:


![image](https://github.com/user-attachments/assets/51f39f38-6ee9-4cb1-9eca-d54789133666)

YanKE01 commented:
> > > @YanKE01 It was already fixed locally yesterday, when.... I noticed this while I was trying to add the Sensor Hub Functionality and also add a unit test for that... however, I'm facing a bit of problem with that.. would you like to help me with that?
> > > or
> > > can I just add the Sensor Hub relevant code and update it as it is, without the unit test?
> > 
> > 
> > What specific issues have you encountered?
> 
> Rather than using sensor_hub, I decided to use the humiture functions. & This is the issue:
> 
> ![image](https://private-user-images.githubusercontent.com/184833829/446927358-51f39f38-6ee9-4cb1-9eca-d54789133666.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc5ODkyMTEsIm5iZiI6MTc0Nzk4ODkxMSwicGF0aCI6Ii8xODQ4MzM4MjkvNDQ2OTI3MzU4LTUxZjM5ZjM4LTZlZTktNGNiMS05ZWNhLWQ1NDc4OTEzMzY2Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyM1QwODI4MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZjE0MTQyMjcwN2QwOWFiMzZjNzI1ZDA4NTFlNWE1OTdjZjIwODE4NWQxMmZiOTUwMWQzYmY5NTg0OGU1Yjc0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.m3bscyGCD2v5TvL3RsWtYP8pDizKXeA3kRc9dwqxp9w)

I think you should add the dependency for `sensor_hub`; you can refer to the `idf_component.yml` of SHT3X:
```yml
version: "0.2.0"
description: I2C driver for SHT3X humiture sensor
url: https://github.com/espressif/esp-iot-solution/tree/master/components/sensors/humiture/sht3x
repository: https://github.com/espressif/esp-iot-solution.git
documentation: https://docs.espressif.com/projects/esp-iot-solution/en/latest/sensors/humiture.html
issues: https://github.com/espressif/esp-iot-solution/issues
dependencies:
  i2c_bus:
    public: true
    override_path: "../../../i2c_bus"
  sensor_hub:
    public: true
    override_path: "../../sensor_hub"
  idf: '>=4.4'
  cmake_utilities: "0.*"

```

jeetrohan commented:
> I think you should add the dependency for `sensor_hub`; you can refer to the `idf_component.yml` of SHT3X:
> 
> ```yaml
> version: "0.2.0"
> description: I2C driver for SHT3X humiture sensor
> url: https://github.com/espressif/esp-iot-solution/tree/master/components/sensors/humiture/sht3x
> repository: https://github.com/espressif/esp-iot-solution.git
> documentation: https://docs.espressif.com/projects/esp-iot-solution/en/latest/sensors/humiture.html
> issues: https://github.com/espressif/esp-iot-solution/issues
> dependencies:
>   i2c_bus:
>     public: true
>     override_path: "../../../i2c_bus"
>   sensor_hub:
>     public: true
>     override_path: "../../sensor_hub"
>   idf: '>=4.4'
>   cmake_utilities: "0.*"
> ```

@YanKE01 
It was already like that

idf._component.yml:

![image](https://github.com/user-attachments/assets/812c6799-d2a5-4534-a647-4544cb47d54f)

I think the problem is here:

![image](https://github.com/user-attachments/assets/6fed93f9-5585-497c-89b4-13c9c61b4092)

& in particular with this part:

![image](https://github.com/user-attachments/assets/45a04b15-d2a4-4054-a582-d81195aacab6)

There are no proper instructions, what should be here:
    target_link_libraries(${COMPONENT_LIB} INTERFACE "-u **humiture_aht20_init"**)

YanKE01 commented:
> > I think you should add the dependency for `sensor_hub`; you can refer to the `idf_component.yml` of SHT3X:
> > ```yaml
> > version: "0.2.0"
> > description: I2C driver for SHT3X humiture sensor
> > url: https://github.com/espressif/esp-iot-solution/tree/master/components/sensors/humiture/sht3x
> > repository: https://github.com/espressif/esp-iot-solution.git
> > documentation: https://docs.espressif.com/projects/esp-iot-solution/en/latest/sensors/humiture.html
> > issues: https://github.com/espressif/esp-iot-solution/issues
> > dependencies:
> >   i2c_bus:
> >     public: true
> >     override_path: "../../../i2c_bus"
> >   sensor_hub:
> >     public: true
> >     override_path: "../../sensor_hub"
> >   idf: '>=4.4'
> >   cmake_utilities: "0.*"
> > ```
> 
> @YanKE01 It was already like that
> 
> ah20 idf._component.yml:
> 
> ```
> 
> name: aht20
> 
> version: "1.1.0"
> 
> author: Rohan Jeet <jeetrohan92@gmail.com>
> 
> description: AHT20 Temperature and Humidity Sensor I2C Driver for ESP-IDF
> 
> issues: https://github.com/espressif/esp-iot-solution/issues
> 
> repository: git://github.com/espressif/esp-iot-solution.git
> 
> url: https://github.com/espressif/esp-iot-solution/tree/master/components/sensors/humiture/aht20
> 
> dependencies:
> 
>   i2c_bus:
>     public: true
>     override_path: "../../../i2c_bus"
>   
>   sensor_hub:
>     public: true
>     override_path: "../../sensor_hub"
> 
>   idf: ">=4.0"
> ```
> 
> I think the problem is here, aht20 CMakeLists.txt:
> 
> ```
> idf_component_register(SRCS "aht20.c"
>                     INCLUDE_DIRS "include"
>                     REQUIRES "sensor_hub")
>                     
> 
> if(CONFIG_SENSOR_INCLUDED_HUMITURE)
>     target_link_libraries(${COMPONENT_LIB} INTERFACE "-u humiture_aht20_init")
> endif()
> 
> include(package_manager)
> cu_pkg_define_version(${CMAKE_CURRENT_LIST_DIR})`
> 
> 
> & in particular with this part:
> 
> `if(CONFIG_SENSOR_INCLUDED_HUMITURE)
>     target_link_libraries(${COMPONENT_LIB} INTERFACE "-u humiture_aht20_init")
> endif()
> ```
> 
> There are no proper instructions, what should be here: `target_link_libraries(${COMPONENT_LIB} INTERFACE "-u humiture_aht20_init")` in particular this: **-u humiture_aht20_init**
> 
> and if you will read the error: it mentions about idf_component_register in Cmakelists.txt file not idf_component.yml
> 
> This is test app CMakelists.txt: `idf_component_register(SRCS "aht20_Sensor_Hub_Humiture.c" INCLUDE_DIRS "." PRIV_REQUIRES unity test_utils aht20 sensor_hub)`
> 
> & its idf_component.yml : `dependencies: aht20: version: "*" override_path: "../../../" sensor_hub: public: true override_path: "../../../../../sensor_hub"`

Perhaps support for the sensor hub can be added in a separate PR next time, as this PR already includes too many changes.

jeetrohan commented:
> > > I think you should add the dependency for `sensor_hub`; you can refer to the `idf_component.yml` of SHT3X:
> > > ```yaml
> > > version: "0.2.0"
> > > description: I2C driver for SHT3X humiture sensor
> > > url: https://github.com/espressif/esp-iot-solution/tree/master/components/sensors/humiture/sht3x
> > > repository: https://github.com/espressif/esp-iot-solution.git
> > > documentation: https://docs.espressif.com/projects/esp-iot-solution/en/latest/sensors/humiture.html
> > > issues: https://github.com/espressif/esp-iot-solution/issues
> > > dependencies:
> > >   i2c_bus:
> > >     public: true
> > >     override_path: "../../../i2c_bus"
> > >   sensor_hub:
> > >     public: true
> > >     override_path: "../../sensor_hub"
> > >   idf: '>=4.4'
> > >   cmake_utilities: "0.*"
> > > ```
> > 
> > 
> > @YanKE01 It was already like that
> > ah20 idf._component.yml:
> > ```
> > 
> > name: aht20
> > 
> > version: "1.1.0"
> > 
> > author: Rohan Jeet <jeetrohan92@gmail.com>
> > 
> > description: AHT20 Temperature and Humidity Sensor I2C Driver for ESP-IDF
> > 
> > issues: https://github.com/espressif/esp-iot-solution/issues
> > 
> > repository: git://github.com/espressif/esp-iot-solution.git
> > 
> > url: https://github.com/espressif/esp-iot-solution/tree/master/components/sensors/humiture/aht20
> > 
> > dependencies:
> > 
> >   i2c_bus:
> >     public: true
> >     override_path: "../../../i2c_bus"
> >   
> >   sensor_hub:
> >     public: true
> >     override_path: "../../sensor_hub"
> > 
> >   idf: ">=4.0"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think the problem is here, aht20 CMakeLists.txt:
> > ```
> > idf_component_register(SRCS "aht20.c"
> >                     INCLUDE_DIRS "include"
> >                     REQUIRES "sensor_hub")
> >                     
> > 
> > if(CONFIG_SENSOR_INCLUDED_HUMITURE)
> >     target_link_libraries(${COMPONENT_LIB} INTERFACE "-u humiture_aht20_init")
> > endif()
> > 
> > include(package_manager)
> > cu_pkg_define_version(${CMAKE_CURRENT_LIST_DIR})`
> > 
> > 
> > & in particular with this part:
> > 
> > `if(CONFIG_SENSOR_INCLUDED_HUMITURE)
> >     target_link_libraries(${COMPONENT_LIB} INTERFACE "-u humiture_aht20_init")
> > endif()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There are no proper instructions, what should be here: `target_link_libraries(${COMPONENT_LIB} INTERFACE "-u humiture_aht20_init")` in particular this: **-u humiture_aht20_init**
> > and if you will read the error: it mentions about idf_component_register in Cmakelists.txt file not idf_component.yml
> > This is test app CMakelists.txt: `idf_component_register(SRCS "aht20_Sensor_Hub_Humiture.c" INCLUDE_DIRS "." PRIV_REQUIRES unity test_utils aht20 sensor_hub)`
> > & its idf_component.yml : `dependencies: aht20: version: "*" override_path: "../../../" sensor_hub: public: true override_path: "../../../../../sensor_hub"`
> 
> Perhaps support for the sensor hub can be added in a separate PR next time, as this PR already includes too many changes.

Thanks @YanKE01.
None of this was issue...
The problem was adding the header itself, only the sensor hub header should be used.
Fixed it all...
Added new example for Sensor Hub and a new test app using humiture, because I am facing problems with sensor hub based test app.. 
and I have no intentions of delving into that currently.
I'll be updating it very soon here and make a request for final review.

jeetrohan commented:
Hi @YanKE01 ,
Final update (hopefully is here).
I suppose you can now officially call this as v2.0 or something? at least not v1.1.0.
And I don't know about you, but I'd like to Rest In Peace,  now :sweat_smile: :sweat_smile: .

jeetrohan commented:
Hi @YanKE01,
Please review the final submission.
And I added Sensor Hub support in this PR itself, because... :sweat_smile: lets just finish this...
AHT20 is a simple sensor with not much to do.
The next two submissions will require much of your attention.
Once this is closed I'll open a new issue regarding those, one after previous resolved.. .
I hope we can merge this soon.
Thanks.

YanKE01 commented on components/sensors/humiture/aht20/Kconfig:
```diff
@@ -0,0 +1,18 @@
+	default n
+
+
+    config AHT20_I2C_CLK_SPEED
+        int "I2C clock speed"
+		default 100000
+	range 1 400000
+        help
+            Clock speed used for i2c communication by AHT20 device.
+	    Limited to maximum of 400KHZ.
```
Considering that I2C-related operations are determined by the `i2c_bus`, I believe the config option here is no longer necessary.

jeetrohan replied:
@YanKE01 ignore what I said earlier..
I ran a test myself... I didn't see any inconsistencies or anything. 
Maybe older AHT20 had these problems..
At worst it may be inconsistent in .001 or 0.01 resolution values and if someone wants that much of consistency they can use some other sensor.
I've fixed and tested this too...

YanKE01 commented on components/sensors/humiture/aht20/README.md:
```diff
@@ -1,32 +1,103 @@
+Note : It is recommended to use clock speeds in upper ranges of 100kHz to 200kHz.
+Higher clock speeds may cause occasional data inconsistencies depending on your board layout and wiring.
+
+![image](https://github.com/user-attachments/assets/fc8680fb-1567-477c-92f8-52dd126e6f9d)
```
Just like the previous suggestion regarding `Kconfig`, the config option is not needed here either.

jeetrohan replied:
please resolve the response on previous suggestion. Based on that I'll follow this through.

YanKE01 commented on components/sensors/humiture/aht20/aht20.c:
```diff
@@ -1,160 +1,401 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+/*
+ * @File: aht20.c
+ *
+ * @brief: AHT20 driver function definitions
+ *
+ * @Date: May 2, 2025
+ *
+ * @Author: Rohan Jeet <jeetrohan92@gmail.com>
+ *
+ */
```
It would be better to remove this section to maintain consistency with the current code style of the repository. Other developers can still find you through the Git history.

YanKE01 commented on components/sensors/humiture/aht20/include/aht20.h:
```diff
@@ -1,79 +1,141 @@
-    i2c_port_t  i2c_port;           /*!< I2C port used to connected AHT20 device */
-    uint8_t     i2c_addr;           /*!< I2C address of AHT20 device, can be 0x38 or 0x39 according to A0 pin */
-} aht20_i2c_config_t;
+typedef aht20_dev_config_t *aht20_handle_t;
```
I still find the definition of handle here a bit strange. Typically, we define handle as a `void*`, with the actual structure hidden in the` .c` file, rather than exposing it directly in the `.h` file.

jeetrohan replied:
@YanKE01 People will call this oversight. Maybe self-inflicted bug... 
But please go ahead and call it a blunder, because that's what this is.

Fixed Locally, will update with other changes, once they are resolved.

YanKE01 commented on components/sensors/humiture/aht20/test_apps/aht20_driver_test/CMakeLists.txt:
```diff
@@ -2,7 +2,8 @@
 
-set(EXTRA_COMPONENT_DIRS "$ENV{IDF_PATH}/tools/unit-test-app/components"
-                         "../../aht20")
+set(COMPONENTS main)
+set(EXTRA_COMPONENT_DIRS "$ENV{IDF_PATH}/tools/unit-test-app/components")
+
```
Please keep the style consistent with the `test_apps `of other components.

jeetrohan replied:
@YanKE01 
you mean like this
```
set(EXTRA_COMPONENT_DIRS "$ENV{IDF_PATH}/tools/unit-test-app/components"
                         "../../aht20")
```
right?

if yes..  fixed.

YanKE01 reviewed (commented):
Hi, @jeetrohan .Thank you very much for your contribution and for testing the `sensor_hub`. I have no additional comments after completing this review.

jeetrohan commented:
> Hi, @jeetrohan .Thank you very much for your contribution and for testing the `sensor_hub`. I have no additional comments after completing this review.

@YanKE01 is it now acceptable for the merge?

YanKE01 commented:
> > Hi, @jeetrohan .Thank you very much for your contribution and for testing the `sensor_hub`. I have no additional comments after completing this review.
> 
> @YanKE01 is it now acceptable for the merge?

Still waiting for review from other relevant personnel.

jeetrohan commented:
> > > Hi, @jeetrohan .Thank you very much for your contribution and for testing the `sensor_hub`. I have no additional comments after completing this review.
> > 
> > 
> > @YanKE01 is it now acceptable for the merge?
> 
> We still need to wait for reviews from other relevant personnel. In the meantime, I think it would be good to squash the commit messages — there are too many at the moment.

@YanKE01 
I'll try to squash the commit messages (haven't done it before ever.) and keep only the relevant ones.
Please keep me posted.
Once we have done a few more drivers I'll demonstrate the solution  I mentioned earlier, and why its needed,  in code myself, but first we need to reach there.
